test/test_type_promotion.py

```diff
@@ -227,6 +227,54 @@ def test_result_type(self):
         self.assertEqual(torch.result_type(torch.tensor([1., 1.], dtype=torch.float), 1.), torch.float)
         self.assertEqual(torch.result_type(torch.tensor(1., dtype=torch.float), torch.tensor(1, dtype=torch.double)), torch.double)
 
+    def test_comparison_ops_with_type_promotion(self):
+        value_for_type = {
+            torch.uint8: (1 << 5),
+            torch.int8: (1 << 5),
+            torch.int16: (1 << 10),
+            torch.int32: (1 << 20),
+            torch.int64: (1 << 30),
+            torch.float16: (1 << 10),
+            torch.float32: (1 << 20),
+            torch.float64: (1 << 30)
+        }
+        comparison_ops = [
+            dict(
+                name="lt",
+                out_op=lambda x, y, d: torch.lt(x, y, out=torch.empty(1, dtype=torch.bool, device=d)),
+                ret_op=lambda x, y: torch.lt(x, y),
+                compare_op=lambda x, y: x < y,
+            ),
+        ]
+        device = self.device
+        for op in comparison_ops:
+            for dt1 in torch.testing.get_all_math_dtypes(device):
+                for dt2 in torch.testing.get_all_math_dtypes(device):
+                    val1 = value_for_type[dt1]
+                    val2 = value_for_type[dt2]
+                    t1 = torch.tensor([val1], dtype=dt1, device=device)
+                    t2 = torch.tensor([val2], dtype=dt2, device=device)
+                    expected = torch.tensor([op["compare_op"](val1, val2)], dtype=torch.bool)
+
+                    out_res = op["out_op"](t1, t2, device)
+                    self.assertEqual(out_res, expected)
+                    self.assertTrue(out_res.dtype == torch.bool)
+                    self.assertTrue(t1.dtype == dt1)
+                    self.assertTrue(t2.dtype == dt2)
+
+                    out_res = op["ret_op"](t1, t2)
+                    self.assertEqual(out_res, expected)
+                    self.assertTrue(out_res.dtype == torch.bool)
+                    self.assertTrue(t1.dtype == dt1)
+                    self.assertTrue(t2.dtype == dt2)
+
+    def test_lt_with_type_promotion(self):
+        for dt in torch.testing.get_all_math_dtypes(self.device):
+            actual = torch.tensor([0], dtype=dt, device=self.device) < 0.5
```

Comment:
is this the only test that tests comparison with a non-dimensioned tensor? We have special rules for promotion that differ if one of the operands is zero-dim vs scalar, and the rules are such that if the scalar is of the same category (integral vs float etc), that it doesn't promote.

Reply:
Can you give an example? How does it work for arithmetic ops?

Reply:
Eg `tensor(1,dtype=uint8) + 1000` does not promote, nor does `tensor([1], dtype=uint8) +tensor(1000)`. We only promote within a "kind" when the category is higher (eg if 1000 was a float). 
You can check the docs in tensor_attributes.rst for more details or I can explain more tomorrow. Should be fine, but could probably use a test to make sure whatever overflows etc are handled in the way you expect.

Reply:
I added tests for cases like: torch.tensor([0], dtype=torch.int) < torch.tensor(0.5)

Reply:
What's the expected behavior for `torch.tensor([250], dtype=uint8) < torch.tensor(10000, dtype=int)`

The int tensor won't get promoted so I'd think it should be the same as `< scalar`, which I believe is expected to throw an error? I saw you have a test that confirms that error when it's explicitly a scalar, but I think you're missing one for when it's a zero-dim tensor. I could imagine that they take different code paths. If for some reason we fail to throw the error, I could imagine the zero-dim tensor getting truncated in conversion and ending up with incorrect results.

The full set of 'categories' are dimensioned tensors, zero-dim tensors, and scalars. If we're within a "kind" we don't use type information from lower category operands in promotion.

Reply:
I added a check that if one the tensor is zero-dim tensor, then other one should have same dtype.